ast/src/analyzed/mod.rs

```diff
@@ -1498,28 +1505,45 @@ impl<T> AlgebraicExpression<T> {
     }
 
     /// Returns the degree of the expressions
-    pub fn degree(&self) -> usize {
+    pub fn degree_with_cache(
+        &self,
+        intermediate_definitions: &BTreeMap<AlgebraicReferenceThin, AlgebraicExpression<T>>,
```

Comment:
Not introduced in this PR, but I find the type here a bit odd. If I understand correctly, it is like this because for `pol x = y * z` we generate `x -> y * z` and `x' -> y' * z'`, so that for a reference inside an expression, we can directly look up its definition.

The issue I see is that `AlgebraicReferenceThin` also allows other types of polynomials. I would have expected something like `poly_id.id` as the key here.

Do you remember why this is needed versus storing just the definitions by id and doing the next logic each time we see a reference? i.e, when seeing `x'`, lookup `x`, get `y * z` and next it to get `y' * z'`?

Reply:
I think I agree. One effect of this type is that for example if `pol x = y' * z`, a definition for `x'` does not exist (as it would be applying the next reference twice). But this could be figured out by the caller too...

Reply:
Exactly! Not needed in this PR, though.